fossas/fossa-cli
# Preflight checks permissions

# Overview

As a user, I want CLI to perform pre-flight checks prior to running any command, in regards to my permissions. 

For example, I want CLI to provide me with remediation message, and halt - if I’m using push-only token, but the action requires full token. This applies to other permission related protection, we have on endpoint - such as TeamAccess, ReleaseGroup etc.

Notes

This should account for:

- fossa report

- fossa test

- fossa analyze (when mutating policy, other attr etc. via .fossa.yml after project is creating)

- Other optional functionality:

   - VSI

   - Snippet Scanning

   - Vendor Dependency Scanning

  - Etc …

## Acceptance criteria

- fossa-cli performs,

   - Permission check for any API related action, prior to starting long running compute/IO work. If the user does not have permission, it halts and provides remediation steps.

- fossa-cli does not perform permission related checks if --output flag is used, and SERVER is not communicated.

## Testing plan

PreflightChecksSpec.hs

Manual tests flow:

To trigger InvalidEditProjectPermission:
- Create an org role with project view permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY> ` on a project that has already been scanned

To trigger InvalidCreateProjectPermission:
- Create an org role with project view permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY> ` on a project that has NOT already been scanned

To trigger InvalidCreateTeamProjectPermission:
- Create a team role with project view permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY>  --team <TEAM>` on a project that has NOT already been scanned

To trigger InvalidCreateProjectOnlyToTeamPermission:
- Create a team role with project create permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY> ` on a project that has NOT already been scanned (do not pass a team argument) 

To trigger InvalidEditReleaseGroupPermission:
- Create an org role with release group view permissions and project edit permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY>  --release-group-name <NAME> --release-group-release <RELEASE>` on a project that has NOT already been scanned 

To trigger InvalidCreateTeamProjectsForReleaseGroupPermission:
- Create a team role with release group edit permissions and project edit permissions and run: `fossa analyze -e http://localhost:9578 --fossa-api-key <KEY>  --release-group-name <NAME> --release-group-release <RELEASE> --team <TEAM>` on a project that has NOT already been scanned. This release group should be associated with multiple teams and the team role should only have project edit permissions for ONLY the team that is being passed

## Risks

Wanted confirmation on the structure of `CustomBuildUploadPermissions` to see if that's a suitable response type

## Metrics

## References

- [ANE-1309](https://fossa.atlassian.net/browse/ANE-1309)
- [Core pr](https://github.com/fossas/FOSSA/pull/11992)

## Checklist



[ANE-1309]: https://fossa.atlassian.net/browse/ANE-1309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
